src/VisualStudio/Core/Def/Implementation/ChangeSignature/ChangeSignatureDialogViewModel.cs

```diff
@@ -20,51 +22,114 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.ChangeSignature
+                _disabledParameters.Add(_thisParameter);
             }
 
+            _declarationParts = symbol.ToDisplayParts(s_symbolDeclarationDisplayFormat);
+
+            _parametersWithoutDefaultValues = CreateParameterViewModels(parameters.ParametersWithoutDefaultValues, ref initialDisplayIndex);
+            _parametersWithDefaultValues = CreateParameterViewModels(parameters.RemainingEditableParameters, ref initialDisplayIndex);
+
             if (parameters.ParamsParameter != null)
             {
-                _paramsParameter = new ParameterViewModel(this, parameters.ParamsParameter);
+                _paramsParameter = new ExistingParameterViewModel(this, parameters.ParamsParameter, initialDisplayIndex++);
             }
 
-            _symbol = symbol;
-            _declarationParts = symbol.ToDisplayParts(s_symbolDeclarationDisplayFormat);
+            var selectedIndex = parameters.SelectedIndex;
+            // Currently, we do not support editing the ThisParameter. 
+            // Therefore, if there is such parameter, we should move the selectedIndex.
+            if (parameters.ThisParameter != null && selectedIndex == 0)
+            {
+                // If we have at least one paramter after the ThisParameter, select the first one after This.
+                // Otherwise, do not select anything.
+                if (parameters.ParametersWithoutDefaultValues.Length + parameters.RemainingEditableParameters.Length > 0)
+                {
+                    this.SelectedIndex = 1;
+                }
+                else
+                {
+                    this.SelectedIndex = null;
+                }
+            }
+            else
+            {
+                this.SelectedIndex = selectedIndex;
+            }
+        }
 
-            _parameterGroup1 = parameters.ParametersWithoutDefaultValues.Select(p => new ParameterViewModel(this, p)).ToList();
-            _parameterGroup2 = parameters.RemainingEditableParameters.Select(p => new ParameterViewModel(this, p)).ToList();
+        public AddParameterDialogViewModel CreateAddParameterDialogViewModel()
+            => new AddParameterDialogViewModel(_document, _insertPosition);
 
-            var selectedIndex = parameters.SelectedIndex;
-            this.SelectedIndex = (parameters.ThisParameter != null && selectedIndex == 0) ? 1 : selectedIndex;
+        List<ParameterViewModel> CreateParameterViewModels(ImmutableArray<Parameter> parameters, ref int initialIndex)
+        {
+            var list = new List<ParameterViewModel>();
+            foreach (ExistingParameter existingParameter in parameters)
+            {
+                list.Add(new ExistingParameterViewModel(this, existingParameter, initialIndex));
+                initialIndex++;
+            }
+
+            return list;
         }
 
         public int GetStartingSelectionIndex()
```

Comment:
Does anything actually call this?

Reply:
Yes, when the dialog is initialized (from the method name, not from a parameter/argument location) and we have to decide which parameter to select.